### PR TITLE
feat: new claiming behavior certification on learn

### LINKED ISSUE #40403: Opening a cert on a private profile redirects to the homepage
If I set my profile to private and try to open a cert from the settings page or from the timeline on my portfolio, I get a message that I need to set my profile to public to view a cert and get redirected to the home page. Seems like a user should get redirected to the page they came from.

Or possibly, remove the flash message and redirect in favor of something like we do for the portfolio pages - where I can open a cert with my profile private, but it says something like "Your profile is private, only you can see this certificate. Make your profile public in order for others to see it."

### LINKED ISSUE #41040: edge case in which it should fall-back on username from name in showing cert
If there is no name set in the settings, doesn't matter if name is public or private, instead of falling back to username, it gives an error and certificate is not shown

### LINKED ISSUE #43358: Certification card checks the "Set your name and make it public" requirement incorrectly.
It seems that "Set your name and make it public" check mark check for the name being public but not if the name is set or not.
Adding another checkmark for setting the username and linking it to the related setting section is a plausible solution.

After setting the name to public and clicking on the `claim certification` button displays the following flash message:

Screen Shot 2021-08-31 at 1 32 23 PM

### PR DESCRIPTION
**Pull request by Sembauke**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #43358 
Closes #41040
Closes #40403

### CODE REVIEW

**Comment by Sembauke** [comment]
I am not entirely sure what we agreed on for these issue's, but here is a way to do it.

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 1]
```diff
@@ -89,6 +89,12 @@ const CertChallenge = ({
       )?.show ?? false
     );
 
+    if (!name) {
+      steps.isNameSet = false;
+    } else {
+      steps.isNameSet = true;
+    }
```
Comment (ShaunSHamilton) [inline client/src/templates/Introduction/components/CertChallenge.js:96]:
```suggestion
    steps.isNameSet = !!name;
```

Comment (Sembauke) [reply client/src/templates/Introduction/components/CertChallenge.js]: Is this new syntax?

Comment (nhcarrigan) [reply client/src/templates/Introduction/components/CertChallenge.js]: I don't believe it's that new.

Comment (ShaunSHamilton) [reply client/src/templates/Introduction/components/CertChallenge.js]: @Sembauke We could always make it more fun:
```js
steps.isNameSet = !!!!!!!!!!!!name;
```

_I wonder if there is a performance penalty for doing that 🤔

**Review by naomi-lgbt — DISMISSED** [review]
This is looking good. Thanks for doing this!

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
This does not really close either linked issue. Nor should we be mutating the `steps` object within the component that uses it (i.e. adding the `isNameSet` property). Typescript would be complaining, but Redux is not typed yet.

What is needed to close the issues is:
- Change requirements to claim certification (server-side too) to require a Camper's name be set.

What we need clarification on (although, we might be good to go with) is if a Camper has their `name` set to `private`, is their certification hidden? Or, does it default to their username (current behaviour)?

---

My final thought:
- `name` not set -> Cannot claim certificate (needs to be checked on user record during verification)
- `name === private` -> Certificate hidden (cannot be viewed by anyone) -> Flash message explaining it needs to be made public (same as `certifications === private`)

**Comment by Sembauke** [comment]
Yeah, I wasn't sure either 👍

**Comment by Sembauke** [comment]
I am not entirely sure what you are suggesting here @ShaunSHamilton,

Are you suggesting not adding an extra step and adding a flash message instead?

**Comment by ShaunSHamilton** [comment]
> Are you suggesting not adding an extra step and adding a flash message instead?

I am suggesting two options:
1) Remove the `name` step entirely, because it is not a technical requirement to claim a certificate (we just added it, because Campers were getting confused with their username showing up on their cert)
2) What I mentioned above as the needed logic:
> name not set -> Cannot claim certificate (needs to be checked on user record during verification)
> name === private -> Certificate hidden (cannot be viewed by anyone) -> Flash message explaining it needs to be made public (same as certifications === private)

For `2`, essentially, we are enforcing the requirement for `name` to be set and public. Currently, it is not enforced - a Camper can claim a certificate without doing prior.

Hope this is clear. If not, we are probably best hoping on a call with whoever else wants, and discussing for a finality.

**Comment by Sembauke** [comment]
image

I get this message after removing the steps for claiming the cert.

File: client/i18n/locales/english/translations.json [hunk 2]
```diff
@@ -450,6 +449,7 @@
     "refresh-needed": "You can only use the PaymentRequest button once. Refresh the page to start over.",
     "username-not-found": "We could not find a user with the username \"{{username}}\"",
     "add-name": "This user needs to add their name to their account in order for others to be able to view their certification.",
+    "name-not-public": "This user has their name set to private in order to view their certificaton it needs to be set to public.",
```
Comment (nhcarrigan) [inline client/i18n/locales/english/translations.json:452]:
```suggestion
    "name-not-public": "This user has their name set to private. In order to view their certification it needs to be set to public.",
```

File: client/i18n/locales/english/translations.json [hunk 3]
```diff
@@ -569,7 +569,8 @@
     "intro": "Complete the following steps to claim and view your {{i18nCertText}}",
     "complete-project": "Complete {{i18nCertText}} Projects",
     "accept-honesty": "Accept our Academic Honesty Policy",
-    "set-name": "Set your name, and make it public",
+    "set-name-public": "Set your name settings to public",
```
Comment (nhcarrigan) [inline client/i18n/locales/english/translations.json:572]:
```suggestion
    "set-name-public": "Set your name to be public",
```
tbh I'm not happy with either verbiage but can't think of something better at the moment

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

**Comment by moT01** [comment]
Sorry if I'm late to the conversation here - I was trying to take a look at this, here's some things I noticed:

In this PR, the cert claiming widget shows 4 steps:
```
Complete the following steps to claim and view your Responsive Web Design Certification
-Accept our Academic Honesty Policy
-Complete Responsive Web Design Certification Projects
-Set your profile settings to public
-Set your certification settings to public
```
If a name is required to claim or view a cert, I think it should be included in the steps. I don't think a name should be required to claim a cert since it hasn't been required in the past and you can just set your name to an empty string afterwards. If it's an empty string, uncheck the step. I think the only things required to claim a cert should be the honesty agreement and the required projects. The step for entering your name can be what it was - `Set your name, and make it public`. If there isn't a name set, the link takes you to the place to add a name - if it is set, it can take you to the privacy setting.

The name should have the same message in the privacy settings as the profile and certification:
`Your certifications will be disabled, if set to private.`

We should not fall back to usernames if a name is private. Names need to be public to view a cert.

My opinion anyway. Seems like that would close both the issues.

**Comment by Sembauke** [comment]
Hey @ShaunSHamilton,
I would like to hear your opinion on this above message.

**Comment by ShaunSHamilton** [comment]
I suggest we organise a meeting with Quincy, and hash out the future for the privacy settings. Until then, this PR should be blocked.

**Comment by Sembauke** [comment]
I was not at the meeting, but I got a lot from the notes. Let me know what I missed.

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
This should not happen (ignore the backlight widget 🤦‍♂️ ):
image

The discussed behaviour is
- Campers can **always** view their certs, but a message similar to `profile === private` is given:
image

**Pull request closed without merging**